compiler/circle-mpqsolver/src/ErrorApproximator.cpp

```diff
@@ -0,0 +1,318 @@
+ * W * q_err(A), q_err - quantization error, W - weight matrix, A - activations from previous layer,
+ * so quantization error of matrix multiplication is A * q_err(W) + W * q_err(A).
+ * Estimating its largest value we get
+ * ~ number_of_additions * (A_max * (W_max - W_min) / 255 + W_max * (A_max - A_min) / 255)
+ * The following code tries to get total error for quantizing convolution node into Q8.
+ * It's just an heuristic (Metric sensitivity depends highly on derivatives as well).
+ */
+float ErrorApproximator::approximate(CircleNode *node)
+{
+  auto circle_node = loco::must_cast<luci::CircleNode *>(node);
+  auto name = circle_node->name();
+  CircleConst *weights = nullptr;
+  uint32_t volume = 0;
+  std::vector<uint32_t> cir_shape;
+  if (getShape(circle_node, cir_shape) && cir_shape.size() == 4)
+  {
+    volume = cir_shape[1] * cir_shape[2] * cir_shape[0];
+  }
+
+  if (auto conv = dynamic_cast<luci::CircleConv2D *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(conv->filter());
+  }
+  else if (auto dconv = dynamic_cast<luci::CircleDepthwiseConv2D *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(dconv->filter());
+  }
+  else if (auto tconv = dynamic_cast<luci::CircleTransposeConv *>(circle_node))
+  {
+    weights = loco::must_cast<luci::CircleConst *>(tconv->filter());
+  }
+
+  if (weights != nullptr)
+  {
+    float ci_min = 0.f;
+    float ci_max = 0.f;
+
+    auto preds = loco::preds(node);
+    for (const auto &pred : preds)
+    {
+      auto parent_node = loco::must_cast<const luci::CircleNode *>(pred);
+      if (has_min_max(parent_node))
+      {
+        auto quantparam = parent_node->quantparam();
+        if (quantparam->min.size() > 0)
+        {
+          ci_min = quantparam->min[0];
+          ci_max = quantparam->max[0];
+        }
+      }
+    }
+
+    // channel-size min, max
+    std::vector<float> min_values;
+    std::vector<float> max_values;
+    {
+      int32_t channel_dim_index = 0;
+      cal_minmax_per_channel(weights, min_values, max_values, channel_dim_index);
+    }
+
+    // ranges (max_values = min_values)
+    std::vector<float> ranges = max_values;
+    for (uint32_t i = 0; i < min_values.size(); ++i)
+    {
+      ranges[i] -= min_values[i];
+    }
+
+    // maximal weight value across all channels
+    float w_max = 0;
+    {
+      for (int i = 0; i < (int)max_values.size(); ++i)
+      {
+        w_max = std::max(w_max, (float)::fabs(max_values[i]));
+        w_max = std::max(w_max, (float)::fabs(min_values[i]));
+      }
+    }
+
+    // total weight quantization error across all channels
+    // so maximal error of quantization is ~ (max_value - min_value) / 255
+    // omitting 255 term we get that maximal error of quantization is just its range
+    float sum_err = 0.f;
+    for (auto cur_err : ranges)
+    {
+      sum_err += cur_err;
+    }
+
+    // maximal quantization error introduced by quantizimh weights
```

Comment:
Or "introduced by weight quantization"?
```suggestion
    // maximal quantization error introduced by quantizing weights
```